### PR TITLE
[CI] Make bazel sharding for parallel buildkite more intelligent

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

This PR implements two changes to our `bazel-sharding.py` script, used for determining which bazel tests to run on each instance when buildkite parallelism is used:
* An ability to filter tests before they are sharded, using the same logic as `bazel test`. This is done by specifying the `--tag_filters` argument, eg. `--tag_filters=air,-gpu`. If we filter tests with `bazel test` *after* they are sharded, we can end up with imbalanced shards as eg. all tests we want to filter out are assigned to one shard. This feature is enabled for Serve tests and it will be required for the changes I want to make to AIR CI.
* A new algorithm to balance the shards, finally implementing what that comment was asking for all this time. Instead of trying to assign the same number of tests (which have variable timeouts) to each shard, the new algorithm tries to make sure each shard will finish in more or less the same time. This is achieved through a simple but good enough heuristic. The old algorithm can still be accessed through the `--sharding_strategy` argument.

Those two changes do cause the complexity of the script to increase, necessitating proper testing. In order to facilitate that, this PR also adds a basic buildkite test harness for CI tools/scripts.

After this PR is merged, the next step will be to move most of our manually parallelized jobs to use buildkite parallelism with the new logic here.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .buildkite/pipeline.build.yml [hunk 1]
```diff
@@ -591,3 +592,12 @@
       python/ray/tests/lightgbm/...
       python/ray/tests/ml_py36_compat/...
       python/ray/tests/xgboost/...
+
+
+- label: ":toolbox: CI Tools"
+  conditions: ["RAY_CI_TOOLS_AFFECTED"]
```
Comment (krfricke) [inline .buildkite/pipeline.build.yml:598]:
```suggestion
  conditions: ["RAY_CI_TOOLS_AFFECTED"]
  instance_size: small
```

Also, can we move this to `pipeline.test.yml` or do we need toolchains here?

File: ci/run/bazel_sharding/bazel_sharding.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            for i, shard in enumerate(shards)
+        ),
+        file=sys.stderr,
+    )
+    return sorted([rule.name for rule in shards[index]])
+
+
+def main(
+    targets: List[str],
+    index: int,
```
Comment (krfricke) [inline ci/run/bazel_sharding/bazel_sharding.py:292]:
```suggestion
    targets: List[str],
    *,
    index: int,
```

Can we force kwargs here so we end up with nicer calls?

File: ci/run/bazel_sharding/tests/test_bazel_sharding.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        os.path.join(tmpdir, WORKSPACE_KEY, "BUILD"),
+    )
+    cwd = os.getcwd()
+    os.chdir(os.path.join(tmpdir, WORKSPACE_KEY))
+    yield
+    os.chdir(cwd)
+    shutil.rmtree(tmpdir, ignore_errors=True)
+
+
+def test_bazel_sharding_end_to_end(mock_build_dir):
```
Comment (krfricke) [inline ci/run/bazel_sharding/tests/test_bazel_sharding.py:50]: Can you add docstrings to all these tests explaining what the test is doing and what we assert?

**Review by krfricke — CHANGES_REQUESTED** [review]
Generally this is great, a few highlevel comments:

- The code requires more comments and documentation. It's hard to see what is happening, and we want to make it easy for future maintainers to understand the code quickly.
- You don't need to fully document code that you didn't change, but if you have any pointers you can quickly add that would be helpful
- For structure, let's create a `ci/tests` directory where we put the BUILD file and a subdirectory for test bazel sharding test (so it contains the mock buildfile)
- Let's either keep the original bazel-sharding.py file or remove the symlink

I'll do a closer review of the new algorithm once there are more comments so I can dive into this a bit easier.

Thanks!

**Comment by Yard1** [comment]
In terms of structure, I tried to make the `ci/tests` directory and realized it is not possible, because:
* All the tested files (so `bazel_sharding.py` in this case) have to be in subdirectories from where the BUILD file is present. Therefore, the BUILD file would need to be in `ci`
* When I put the BUILD file in `ci`, it broke multiple entries in other BUILD files which operate under the assumption that `ci` is not a bazel subpackage

I imagine that could all be fixed eventually but I didn't want to make multiple seemingly unrelated changes in this PR, therefore I opted for the structure shown here. We can either refactor that in a followup, or try to make it work right now but I think it would be best to leave it for later.

**Comment by Yard1** [comment]
I added some comments, PTAL @krfricke

**Comment by krfricke** [comment]
> Why would we want to keep the old file btw? If someone wants the older one, they can just grab it off github. Having two scripts would be confusing.

I meant instead of the one in the sub directory. Let's just remove the symlink then - taking another look now!

File: ci/run/bazel_sharding/bazel_sharding.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+        """
+        Get the next longest (taking up the most time) BazelRule from the
+        ``rules_grouped_by_time`` list.
+        """
+        item = None
+        for _, items in rules_grouped_by_time:
+            if items:
+                item = items.pop()
+                break
+        return item
```
Comment (krfricke) [inline ci/run/bazel_sharding/bazel_sharding.py:276]:
```suggestion
                return = items.pop()
```

Comment (krfricke) [reply ci/run/bazel_sharding/bazel_sharding.py:276]: I guess alternatively we could just use a max heap

File: ci/run/bazel_sharding/bazel_sharding.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+
+    def add_rule_to_best_shard(rule_to_add: BazelRule):
```
Comment (krfricke) [inline ci/run/bazel_sharding/bazel_sharding.py:278]: This looks good to me, but maybe we can move this out of the local scope and test it in the unit tests

**Review by krfricke — COMMENTED** [review]
Thank you very much for the documentation, it helps me a lot to understand what is going on there.

The code and algorithm look good to me. I think we could unit test parts of the algorithm more (e.g. `add_rule_to_best_shard()`, or `rule.actual_timeout_s`), as we currently only have end to end tests for this.

File: ci/run/bazel_sharding/bazel_sharding.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+        if target.startswith("-"):
+            excluded_targets.append(target[1:])
+        else:
+            included_targets.append(target)
+    return included_targets, excluded_targets
+
+
+def quote_targets(targets: Iterable[str]) -> str:
+    """Quote each target in a list so that it can be passed used in subprocess."""
+    return (" ".join("'{}'".format(t) for t in targets)) if targets else ""
```
[the remark this answers is not in the document]
Comment (Yard1) [reply ci/run/bazel_sharding/bazel_sharding.py:112]: It was there before, but we may as well use this opportunity to refactor!

File: ci/run/bazel_sharding/bazel_sharding.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+
+
+def split_tag_filters(tag_str: str) -> Tuple[Set[str], Set[str]]:
```
[the remark this answers is not in the document]
Comment (Yard1) [reply ci/run/bazel_sharding/bazel_sharding.py:115]: Correct

**Review by krfricke — APPROVED** [review]
Awesome!

**Review by krfricke — APPROVED** [review]

**Pull request merged**